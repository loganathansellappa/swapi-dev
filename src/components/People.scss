.PeopleContainer {
  display: flex;
  flex-direction: column;
  height: 100%;
  padding: 10px;
  .searchInput {
    width: 100%;
    background-image: linear-gradient(0deg, red 0%, blue 100%);
    color: #ffffff;
    font-size: 2rem;
    padding: 12px 20px;
    margin: 8px 0;
    display: inline-block;
    border: 1px solid #ccc;
    border-radius: 4px;
    box-sizing: border-box;
  }
  .people {
    display: flex;
    flex-wrap: wrap;
    padding-left: 10%;
    padding-top: 1%;
    padding-right: 10%;
    justify-content: center;
  }
  .paginate {
    display: flex;
    flex-wrap: nowrap;
    justify-content: center;
    margin-top: 10%;
  }
}
