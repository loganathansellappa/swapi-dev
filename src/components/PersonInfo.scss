.container {
  background-image: linear-gradient(260deg, #2376ae 0%, #c16ecf 100%);
  color: #FFFFFF;
  min-height: 100vh;
  .topLinks {
    display: flex;
    justify-content: space-around;
    button {
      padding: 1rem 4rem;
      font-size: 2rem;
      margin: 2rem 0;
      margin-right: 1rem;
      border-radius: 15px;
    }
  }

  .personDetail {
    border-radius: 0.5rem;
    padding: 1%;
    margin: 1%;
    height: 100%;
    .propContainer {
      display: flex;
      font-size: 2rem;
      height: 100%;

      .label {

        font-weight: bold;
        min-width: 500px;
      }
      .value {
        margin-left: 1%;
        flex-wrap: wrap;
        height: 100%;
        display: flex;
      }
    }
    .personInfo {
      display: flex;
      flex-direction: column;
    }

    .favourite {
      align-self: center;
    }

    .favourite:hover {
      cursor: pointer;
    }

    .resourceLink {
      text-decoration: none;
      color: yellow;
      min-width: 0;
      margin-right: 3%;
      cursor: pointer;

    }

    .resourceLink:hover {
      border-bottom: 7px solid #FFFFFF;
      transform: translate3d(0, 0, 0);
      backface-visibility: hidden;
      perspective: 1000px;
      color: #FFFFFF;
    }
  }
}
