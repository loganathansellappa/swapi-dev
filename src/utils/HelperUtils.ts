export const apiService = () => `${process.env.SWAP_URL}`;
export const capitalize = (s = 'n/a') => s[0].toUpperCase() + s.slice(1);
