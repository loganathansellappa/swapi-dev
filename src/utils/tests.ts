import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { BrowserRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';

export const renderWithRouter = (ui: JSX.Element, { route = '/' } = {}) => {
  window.history.pushState({}, 'Test page', route);

  return {
    user: userEvent.setup(),
    ...render(ui, { wrapper: BrowserRouter }),
  };
};
export const testStore = () => {
  const mockStore = configureStore();
  const initialState = { favoritePeople: { favourites: [] } };
  const store = mockStore(initialState);
  return store;
};

export const mockResource = () => {
  return {
    name: 'Luke Skywalker',
    height: '172',
    mass: '77',
    hair_color: 'blond',
    skin_color: 'fair',
    eye_color: 'blue',
    birth_year: '19BBY',
    gender: 'male',
    homeworld: 'https://swapi.dev/api/planets/1/',
    films: [
      'https://swapi.dev/api/films/1/',
      'https://swapi.dev/api/films/2/',
      'https://swapi.dev/api/films/3/',
      'https://swapi.dev/api/films/6/',
    ],
    species: [],
    vehicles: [
      'https://swapi.dev/api/vehicles/14/',
      'https://swapi.dev/api/vehicles/30/',
    ],
    starships: [
      'https://swapi.dev/api/starships/12/',
      'https://swapi.dev/api/starships/22/',
    ],
    created: '2014-12-09T13:50:51.644000Z',
    edited: '2014-12-20T21:17:56.891000Z',
    url: 'https://swapi.dev/api/people/1/',
  };
};
