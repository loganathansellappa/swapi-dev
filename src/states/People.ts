import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { PersonType, ResourceType } from '../@types/People';

export interface PeopleState {
  favourites: Array<ResourceType> | [];
}

const initialState: PeopleState = {
  favourites: [],
};

export const peopleSlice = createSlice({
  name: 'people',
  initialState,
  reducers: {
    add: (state, action: PayloadAction<ResourceType>) => {
      state.favourites = [...state.favourites, action.payload];
    },
    remove: (state, action: PayloadAction<string>) => {
      state.favourites = state.favourites.filter(
        (person) => action.payload !== person.url
      );
    },
  },
});

export const { add, remove } = peopleSlice.actions;

export default peopleSlice.reducer;
