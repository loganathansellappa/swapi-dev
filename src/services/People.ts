import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react'
import {ResourceResponse, ResourceType} from "../@types/People";
import {apiService} from "../utils/HelperUtils";

export const peopleApi = createApi({
    reducerPath: 'peopleApi',
    baseQuery: fetchBaseQuery({ baseUrl: apiService() }),
    endpoints: (builder) => ({
        getResource: builder.query<ResourceResponse, string>({
            query: (id) => `${id}`,
        }),
    }),
})

export const {useGetResourceQuery } = peopleApi