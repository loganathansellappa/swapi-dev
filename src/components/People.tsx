import React, {useMemo, useState} from "react";
import {ResourceResponse, ResourceType} from "../@types/People";
import {Person} from "./Person";
import Error from "./Error";
import {useGetResourceQuery} from '../services/People'
import {Loader} from "./Loader";
import "./People.scss";


const People: React.FunctionComponent  = () => {
    const [page, setPage] = useState<number>(1);
    const [search, setSearch] = useState('');
    const [data, setData] = useState<ResourceResponse | null>(null);
    const [loading, setLoading] = useState(false);
    const [error, setError] = useState(false);

    const updateSearch = (val: string) => {
        setPage(1)
        setSearch(val)
    }

    const peopleData =  useGetResourceQuery(`people?search=${search}&page=${page}`)
    const planetsData =  useGetResourceQuery(`planets?search=${search}&page=${page}`, {skip: !search.length})
    const shipData = useGetResourceQuery(`starships?search=${search}&page=${page}`, {skip: !search.length})

    useMemo(() => {
        let responses: ResourceResponse = {
            count: 0,
            next: null,
            previous: null,
            results: []
        }
        let validData = search.length ?  [peopleData, planetsData, shipData] : [peopleData]
        for (const { currentData:data, error:err, isLoading, isFetching } of validData) {
            console.log(err, 'err')
            if(isLoading || isFetching && !loading) {
                setLoading(true)
            }
            if(err && !error) {
                setLoading(error)
            }
            if(data) {
                if(data?.next) {
                    responses.next = data.next
                }
                setLoading(false)
                setError(false)
                responses.results = [...responses.results, ...data.results]
            }
        }
        setData(responses as ResourceResponse)
    }, [peopleData.data, planetsData.data, shipData.data])

    if (error) {return <Error />}

    return (
        <div className='PeopleContainer'>
            <div className="search">
                <input className={'searchInput'} value={search} placeholder="Search People/Planet/StarShips.." onChange={(e) =>  updateSearch((e.target as HTMLInputElement).value)}/>
            </div>
            {
                loading ?  <Loader size={100} className={'inline'}/> :
                <div className="people">
                    {data?.results?.map((person: ResourceType) => (
                        <Person key={person.url} person={person} />
                    ))}
                </div>
            }
                <div className="paginate">
                    <button  onClick={() => setPage((old) => Math.max(old - 1, 1))}  disabled={page === 1}>
                        {loading ? <Loader size={10} /> : 'Previous'}
                    </button>
                    <button onClick={() => setPage((old) => old + 1)} disabled={!data?.next} >
                        {loading ? <Loader size={10} />: 'Next'}
                    </button>
                </div>
        </div>
    );
};

export default People;
