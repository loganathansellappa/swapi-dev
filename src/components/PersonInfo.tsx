import React, {ReactElement, useCallback, useMemo} from "react";
import {useParams, useNavigate, Link} from 'react-router-dom';
import Error from "./Error";
import {useGetResourceQuery} from "../services/People";
import {capitalize} from "../utils/HelperUtils";
import {Loader} from "./Loader";
import {AnyResourceType, ResourceType} from "../@types/People";
import "./PersonInfo.scss";

export const PersonInfo = () => {
    const { id = 1, type = 'people'} = useParams();
    const navigate = useNavigate();
    const { data: person, error, isLoading, isFetching } = useGetResourceQuery(`${type}/${id}`)

    if (isLoading || isFetching) {return <Loader size={100} />}

    if (error) {return <Error  />}

    const getProperties =(person: ResourceType) => {
        let props = []
        let urls = ['planets', 'species', 'starships', 'vehicles', 'homeworld', 'films', 'people', 'residents', 'characters']
        let peopleType = ['people', 'residents', 'characters']
        for (const key in person) {
            if([...urls, ...peopleType].includes(key) && person[key as keyof ResourceType]) {
                let links = []
                if(Array.isArray(person[key as keyof ResourceType])) {
                    let resourceKey  = peopleType.includes(key) ? 'people' : key
                    person[key as keyof ResourceType].forEach<AnyResourceType>((resourceUrl: string, index: number) =>
                        links.push(<Link className={'resourceLink'} to = {`/resource/${resourceKey}/${resourceUrl.match(/\d/g).join('')}`}>{capitalize(key)}{index+1}</Link>))
                } else {
                    links.push(<Link className={'todo'} to = {`/resource/${key}/${person[key as keyof ResourceType].match(/\d/g).join('')}`} />)
                }
                props.push(<div className="propContainer">
                    <div className='label'>{capitalize(key)}</div>
                    <div className='value'>{links}</div>
                </div>)
            } else {
                props.push(<div className="propContainer">
                    <div className='label'>{capitalize(key)}</div>
                    <div className='value'>{person[key as keyof ResourceType]}</div>
                </div>)
            }
        }
        return props
    }

    return(
        <div className='container'>
            <div className='topLinks'>
                <button  onClick={() =>  navigate(-1)}> Previous Page </button>
                <button onClick={() =>  navigate('/')}>Back to Home</button>
            </div>
            <div className="personDetail">{getProperties(person as unknown as ResourceType)}</div>
        </div>
    )
}