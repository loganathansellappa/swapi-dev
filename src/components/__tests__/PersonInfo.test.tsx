import { mockResource, renderWithRouter, testStore } from '../..//utils/tests';
import { PersonInfo } from '../PersonInfo';
import { Provider } from 'react-redux';
import * as Hooks from '../../services/People';

describe('PersonInfo', () => {
  let element: { container: HTMLElement };

  beforeEach((): void => {
    jest.spyOn(Hooks, 'useGetResourceQuery').mockReturnValue({
      data: mockResource(),
      isError: false,
      isLoading: false,
      refetch: jest.fn(),
    });
    element = renderWithRouter(
      <Provider store={testStore()}>
        <PersonInfo />
      </Provider>
    );
  });

  test('renders PersonInfo component', () => {
    expect(element.container.firstChild).toMatchSnapshot();
  });
});
