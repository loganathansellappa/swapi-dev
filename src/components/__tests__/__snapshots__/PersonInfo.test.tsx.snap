// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonInfo renders PersonInfo component 1`] = `
<div
  class="container"
>
  <div
    class="topLinks"
  >
    <button>
       Previous Page 
    </button>
    <button>
      Back to Home
    </button>
  </div>
  <div
    class="personDetail"
  >
    <div
      class="propContainer"
    >
      <div
        class="label"
      >
        Name
      </div>
      <div
        class="value"
      >
        Luke Skywalker
      </div>
    </div>
    <div
      class="propContainer"
    >
      <div
        class="label"
      >
        Height
      </div>
      <div
        class="value"
      >
        172
      </div>
    </div>
    <div
      class="propContainer"
    >
      <div
        class="label"
      >
        Mass
      </div>
      <div
        class="value"
      >
        77
      </div>
    </div>
    <div
      class="propContainer"
    >
      <div
        class="label"
      >
        Hair_color
      </div>
      <div
        class="value"
      >
        blond
      </div>
    </div>
    <div
      class="propContainer"
    >
      <div
        class="label"
      >
        Skin_color
      </div>
      <div
        class="value"
      >
        fair
      </div>
    </div>
    <div
      class="propContainer"
    >
      <div
        class="label"
      >
        Eye_color
      </div>
      <div
        class="value"
      >
        blue
      </div>
    </div>
    <div
      class="propContainer"
    >
      <div
        class="label"
      >
        Birth_year
      </div>
      <div
        class="value"
      >
        19BBY
      </div>
    </div>
    <div
      class="propContainer"
    >
      <div
        class="label"
      >
        Gender
      </div>
      <div
        class="value"
      >
        male
      </div>
    </div>
    <div
      class="propContainer"
    >
      <div
        class="label"
      >
        Homeworld
      </div>
      <div
        class="value"
      >
        <a
          class="todo"
          href="/resource/homeworld/1"
        />
      </div>
    </div>
    <div
      class="propContainer"
    >
      <div
        class="label"
      >
        Films
      </div>
      <div
        class="value"
      >
        <a
          class="resourceLink"
          href="/resource/films/1"
        >
          Films
          1
        </a>
        <a
          class="resourceLink"
          href="/resource/films/2"
        >
          Films
          2
        </a>
        <a
          class="resourceLink"
          href="/resource/films/3"
        >
          Films
          3
        </a>
        <a
          class="resourceLink"
          href="/resource/films/6"
        >
          Films
          4
        </a>
      </div>
    </div>
    <div
      class="propContainer"
    >
      <div
        class="label"
      >
        Species
      </div>
      <div
        class="value"
      />
    </div>
    <div
      class="propContainer"
    >
      <div
        class="label"
      >
        Vehicles
      </div>
      <div
        class="value"
      >
        <a
          class="resourceLink"
          href="/resource/vehicles/14"
        >
          Vehicles
          1
        </a>
        <a
          class="resourceLink"
          href="/resource/vehicles/30"
        >
          Vehicles
          2
        </a>
      </div>
    </div>
    <div
      class="propContainer"
    >
      <div
        class="label"
      >
        Starships
      </div>
      <div
        class="value"
      >
        <a
          class="resourceLink"
          href="/resource/starships/12"
        >
          Starships
          1
        </a>
        <a
          class="resourceLink"
          href="/resource/starships/22"
        >
          Starships
          2
        </a>
      </div>
    </div>
    <div
      class="propContainer"
    >
      <div
        class="label"
      >
        Created
      </div>
      <div
        class="value"
      >
        2014-12-09T13:50:51.644000Z
      </div>
    </div>
    <div
      class="propContainer"
    >
      <div
        class="label"
      >
        Edited
      </div>
      <div
        class="value"
      >
        2014-12-20T21:17:56.891000Z
      </div>
    </div>
    <div
      class="propContainer"
    >
      <div
        class="label"
      >
        Url
      </div>
      <div
        class="value"
      >
        https://swapi.dev/api/people/1/
      </div>
    </div>
  </div>
</div>
`;
